torch/distributed/_functional_collectives.py

```diff
@@ -166,6 +166,32 @@ def broadcast(self: torch.Tensor, src: int, group: RANK_TYPES, tag: str = ""):
     return _maybe_wrap_tensor(tensor)
 
 
+def scatter(
+    self: torch.Tensor,
+    scatter_list: Optional[List[torch.Tensor]],
+    src: int,
+    group: RANK_TYPES,
+    tag: str = "",
+):
+    """
+    Scatters the list of data tensor to all processes in the given process group.
+
+    Args:
+        self (Tensor): dummy tensor of correct size to receive tensor.
```

Comment:
hmmm I really don't like this `self` to be a dummy tensor that user have to pass. I think think it would force user to allocate an additional tensor just for us to know the correct shape/device 🤔 

What I think could be reasonable "functional collectives":
1. scatter(tensor, scatter_dim, src, ...) -> tensor: where we do the input splitting inside the scatter implementation and then call the c10d version
2. scatter(tensor_list, src, ...) -> tensor: where the input tensor list is required to be passed by every rank from user, and then we just use the `tensor_list[rank]` tensor to correctly allocate the output tensor.

IMO we don't really need to be aligned with the c10d API, as I think the c10d version of scatter is really hard to use (i.e. user have to make these if else branches and explicitly allocate outputs)

Reply:
The discussion I had with @yifuwang is that whether we should defined the `scatter` to be used in a "SPMD" case where every rank should be responsible to allocate input tensors, many of our current c10d APIs have this assumptions, but there're a few exceptions (i.e. alltoall_single)

cc @wconstab if you have some opinions on this

Reply:
Two aspects to this:

* computing the full list on every worker vs rank0
* allocating a full tensor vs just dimension/empty tensor

It does seem like there's _potentially_ benefit to not having to compute the full list on every worker but it does seem more clunky than the broadcast equivalent which uses input as output.

Another option would be to split it into two operators scatter_send and scatter_receive which would be more explicit about the behavior.

Reply:
I think our collectives are mostly following MPI standard under the hood as it's HPC comm, if we look at the MPI scatter semantics https://www.open-mpi.org/doc/v3.1/man3/MPI_Scatter.3.php, MPI basically requiring all rank to call the same API with the same arguments, and the non-root rank would be ignored but user still have to pass them.

IMO I think we should follow the SPMD protocol here and enforce some constraints to user: we should force each process to pass a list or a big tensor, so that every rank would prepare a list of tensor or a single big tensor by user, and then we just use `tensor_list[rank]` to infer the output tensor dtype and device, shape on each rank. Doing a separate dummy tensor arg to infer the output device/dtype is kind of confusing

scatter_list vs single big tensor I don't have big preference, my original intention is that we should follow `reduce_scatter_tensor` semantic, as `reduce_scatter_tensor` is just a reduction + scatter from semantic prospective.

Reply:
> we should follow reduce_scatter_tensor semantic, as reduce_scatter_tensor is just a reduction + scatter from semantic prospective

IMO this might be a bit unnatural. With `reduce_scatter` (ignoring `reduce_scatter_v` for now), all ranks have legit input from which the local output size can be inferred. If `scatter` were to mimic the API, non-source ranks would have to construct dummy inputs just so that the local output size can be inferred. It may also beg the question "what data should these tensors hold?". Generally, I feel using dummy tensor(s) to convey shape and device information is confusing (`empty_like` gets a pass since it has a `like` in its name :) ).

> if we look at the MPI scatter semantics https://www.open-mpi.org/doc/v3.1/man3/MPI_Scatter.3.php

@wanchaol  It's a great idea to look at what MPI does. But if you look closely, even the MPI `scatter` says that the input is "significant only at root". This makes sense because `scatter` is inherently asymmetrical. IMO, instead of trying to make the API feel symmetric, we should focus on only passing essential information into the API.

Reply:
Another point - I don't think we need to force users to use the functional variant if the non-functional variant is easier to use.

The main goal of functional collective is to make collectives PT2 friendly. The functional variant of some collectives happens to be more ergonomic, but it's not always the case. For things like `scatter`, I think it makes sense to recommend users to use `scatter_`. The functional form doesn't need to be pretty. We just need it for functionalization.

For example, most users would use this form: 
```
scatter_(Tensor(a!) output, Tensor[]? inputs) -> Tensor(a!)
```

which get functionalized into this:
```
scatter(Tensors[]? inputs, SymInt[] size, Device device) -> Tensor
```

Reply:
> But if you look closely, even the MPI scatter says that the input is "significant only at root". 
Yeah it's "significant only at root", but it does not mean that it allow user to pass in sth that is None or non-asymmetric, user still need to pass legit inputs on all other non-significant ranks, that's what most HPC API enforces user to.

>This makes sense because scatter is inherently asymmetrical. IMO, instead of trying to make the API feel symmetric, we should focus on only passing essential information into the API.

I disagree actually, if we do this asymmetrical nature, we should also change `broadcast` to follow this semantic, as it's asymmetric as well, we can't just clone instead `broadcast` because other rank could pass "empty/None" thing, and the output of the broadcast on each rank is unknown. I think we should follow `broadcast`, as these two are naturally the same 1-to-many relationship and they should behave the same. 

For other asymmetric APIs like [`send/recv`](https://pytorch.org/docs/stable/distributed.html#torch.distributed.send), there's implicit assumption already baked in there that a send/recv pair must have the same tensor with the same shape/device/dtype, if you want fully asymmetric api, we should also change that to pass in additional `shape/device/dtype` as arguments to send/recv calls

Reply:
> I disagree actually, if we do this asymmetrical nature, we should also change broadcast to follow this semantic, as it's asymmetric as well, we can't just clone instead broadcast because other rank could pass "empty/None" thing, and the output of the broadcast on each rank is unknown.

I actually think the functional variant of broadcast suffers the same problem. If I were to design it now, I would probably not use a dummy tensor to convey the output size and device for non-src ranks.